### PR TITLE
Basic JaxPEAKVI class

### PR DESCRIPTION
**Pull request by RK900**
Partially addresses #1578

### CODE REVIEW

**Review by jjhong922 — CHANGES_REQUESTED** [review]
good start, but need to be very sure the model is implemented in parallel with the torch implementation. Implementing get_latent should help test this.

File: scvi/module/_jaxpeakvae.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        tensors,
+        inference_outputs,
+        generative_outputs,
+        kl_weight: float = 1.0,
+    ):
+        x = tensors[REGISTRY_KEYS.X_KEY]
+        px = generative_outputs["px"]
+        qz = inference_outputs["qz"]
+
+        reconstruction_loss = -px.log_prob(x).sum(-1)
```
[the remark this answers is not in the document]
Comment (RK900) [reply scvi/module/_jaxpeakvae.py:95]: https://optax.readthedocs.io/en/latest/api.html?highlight=binary#optax.sigmoid_binary_cross_entropy

Do we want to use optax for these sorts of things?

[the remark this answers is not in the document]
Comment (jjhong922) [reply scvi/module/_jaxpeakvae.py:95]: that's fine as long as the sign and normalization match torch's

**Comment by RK900** [comment]
Screen Shot 2022-07-28 at 8 08 49 PM

Latent space from JPVI, matching the one from the [PeakVI tutorial](https://docs.scvi-tools.org/en/0.15.1/tutorials/notebooks/PeakVI.html)

File: scvi/module/_jaxpeakvae.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    dropout_rate: float
+    n_hidden: int
+    n_output: int = None
+
+    def setup(self):
+        self.dense1 = Dense(self.n_hidden)
+        self.dense2 = Dense(self.n_hidden)
+        self.dense3 = Dense(self.n_hidden)
+        self.dense4 = Dense(self.n_hidden)
+        if self.n_output is None:
```
Comment (jjhong922) [inline scvi/module/_jaxpeakvae.py:62]: do we need n_input if it's inferred? Seems like we just need n_output to be defined for every instance

File: scvi/module/_jaxpeakvae.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            self.dense5 = Dense(self.n_input)
+        else:
+            self.dense5 = Dense(self.n_output)
+
+        self.batchnorm1 = nn.BatchNorm()
+        self.batchnorm2 = nn.BatchNorm()
+        self.dropout1 = nn.Dropout(self.dropout_rate)
```
Comment (jjhong922) [inline scvi/module/_jaxpeakvae.py:69]: dropout is deterministically set to 0 in peakvae, we can just remove this? same with dropout2

File: scvi/module/_jaxpeakvae.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+class FlaxEncoder(nn.Module):
+    n_input: int
+    n_latent: int
+    n_hidden: int
+    dropout_rate: int
+
+    def setup(self):
+        self.dense1 = Dense(self.n_hidden)
+        self.dense2 = Dense(self.n_hidden)
+        self.dense3 = Dense(self.n_latent)
```
Comment (jjhong922) [inline scvi/module/_jaxpeakvae.py:23]: provide more descriptive names given that these are not sequential like the names would suggest otherwise

File: scvi/module/_jaxpeakvae.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    n_input: int
+    dropout_rate: float
+    n_hidden: int
+    n_output: int = None
+
+    def setup(self):
+        self.dense1 = Dense(self.n_hidden)
+        self.dense2 = Dense(self.n_hidden)
+        self.dense3 = Dense(self.n_hidden)
+        self.dense4 = Dense(self.n_hidden)
```
Comment (jjhong922) [inline scvi/module/_jaxpeakvae.py:61]: i think this should only be 2 layers by default

File: scvi/module/_jaxpeakvae.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+        input_dict = dict(x=x, batch_index=batch_index)
+        return input_dict
+
+    def inference(self, x: jnp.ndarray, batch_index, n_samples: int = 1) -> dict:
+        mean, var = self.encoder(x, training=self.training)
+        stddev = jnp.sqrt(var) + self.eps
+
+        batch = jax.nn.one_hot(batch_index, self.n_batch).squeeze(-2)
+        d = self.d_encoder(x, batch, training=self.training)
```
Comment (jjhong922) [inline scvi/module/_jaxpeakvae.py:145]: can we make this an optional encoder with the `model_depth` kwarg

File: tests/models/test_models.py [hunk 7]
```diff
@@ -87,6 +88,23 @@
 }
 
 
+def test_jax_peakvi():
+    n_latent = 5
+
+    adata = synthetic_iid()
```
Comment (jjhong922) [inline tests/models/test_models.py:94]: use fixture

**Review by jjhong922 — COMMENTED** [review]
latent space is looking good! some comments on more discrepancies with the torch implementation. Doesn't have to have full functionality, but the least we can do is get the defaults right. Another necessary check is to ensure the elbo values are in the same range as the torch implementation.

File: scvi/module/_jaxpeakvae.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+import jax
+import jax.numpy as jnp
+import numpyro.distributions as dist
+from flax import linen as nn
+
+from scvi import REGISTRY_KEYS
+from scvi.module.base import JaxBaseModuleClass, LossRecorder
+
+from ._jaxvae import Dense
```
[the remark this answers is not in the document]
Comment (RK900) [reply scvi/module/_jaxpeakvae.py:11]: Moved it to _base_components

File: tests/models/test_models.py (hunk 7 above)
Comment (RK900) [reply tests/models/test_models.py:94]: I'm gonna refactor these tests in a separate PR, I think I can add fixtures for the scvi tests as well

File: scvi/module/_jaxpeakvae.py (hunk 3 above)
Comment (RK900) [reply scvi/module/_jaxpeakvae.py:69]: In peakvi the default dropout is 0.1

File: scvi/module/_jaxpeakvae.py (hunk 5 above)
Comment (RK900) [reply scvi/module/_jaxpeakvae.py:61]: I think we need all of these layers, one for x and one for batch

File: scvi/module/_jaxpeakvae.py (hunk 3 above)
Comment (jjhong922) [reply scvi/module/_jaxpeakvae.py:70]: This is only true in the encoder, for the decoder its set to 0: https://github.com/scverse/scvi-tools/blob/master/scvi/module/_peakvae.py#L62
